File: src/main/java/codes/qdbp/serverplugin/misc/SleepForward.java
package codes.qdbp.serverplugin.misc;

import org.bukkit.Statistic;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.entity.Pose;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.List;
import java.util.stream.Collectors;

public class SleepForward extends BukkitRunnable{
    private final Player player;

    //for SkipNight command
    private final boolean wasCommand;


    public SleepForward(Player player, boolean wasCommand) {
        this.player = player;
        this.wasCommand = wasCommand;
    }


    public List<Player> getSleepingPlayers(World world) {
        return world.getPlayers().stream()
                .filter(player -> player.getPose() == Pose.SLEEPING)
                .collect(Collectors.toList());
    }

    @Override
    public void run() {
        if (!(13000 < player.getWorld().getTime())) {
            player.setStatistic(Statistic.TIME_SINCE_REST, 0);
            if (!(player.getWorld().isClearWeather())) player.getWorld().setClearWeatherDuration(1);
            cancel();
            return;
        } else if (player.getWorld().isThundering()) {
            if (player.getWorld().getTime() > 23800) {
                player.setStatistic(Statistic.TIME_SINCE_REST, 0);
                if (!(player.getWorld().isClearWeather())) player.getWorld().setClearWeatherDuration(1);
                cancel();
                return;
            }
        } else if (getSleepingPlayers(player.getWorld()).size() == 0 && !wasCommand) {
            cancel();
            return;
        }
        if (player.getWorld().getTime() < 23940) player.getWorld().setTime(player.getWorld().getTime() + 50);
    }
}

File: src/main/kotlin/codes/qdbp/serverplugin/misc/SleepForward.kt
package codes.qdbp.serverplugin.misc

import codes.qdbp.serverplugin.Serverplugin
import org.bukkit.Statistic
import org.bukkit.entity.Player
import org.bukkit.entity.Pose
import org.bukkit.scheduler.BukkitRunnable

fun Player.sleepForward(plugin: Serverplugin, byCommand: Boolean) {

    object : BukkitRunnable() {
        val player = this@sleepForward

        override fun run() {
            val sleepingPlayers = player.world.players.filter { it.pose == Pose.SLEEPING }
            if (13000 >= player.world.time) {
                player.setStatistic(Statistic.TIME_SINCE_REST, 0)
                if (!(player.world.isClearWeather)) player.world.clearWeatherDuration = 1
                cancel()
                return
            } else if (player.world.isThundering) {
                if (player.world.time > 23800) {
                    player.setStatistic(Statistic.TIME_SINCE_REST, 0)
                    if (!(player.world.isClearWeather)) player.world.clearWeatherDuration = 1
                    cancel()
                    return
                }
            } else if (sleepingPlayers.isEmpty() && !byCommand) {
                cancel()
                return
            }

            if (player.world.time < 23940) player.world.time += 50
        }
    }.runTaskTimer(plugin, 0, 0)
}